sympy/assumptions/ask.py

```diff
@@ -1211,6 +1213,30 @@ def _extract_facts(expr, symbol, check_reversed_rel=True):
         return expr.func(*args)
 
 
+def _extract_all_facts(expr, symbol):
+    facts = []
+    if isinstance(symbol, Relational):
+        symbols = (symbol, symbol.reversed)
+    else:
+        symbols = (symbol,)
+    for clause in expr.clauses:
+        args = []
+        for literal in clause:
+            if isinstance(literal.lit, AppliedPredicate):
+                if literal.lit.arg in symbols:
+                    args.append(Literal(literal.lit.func, literal.is_Not))
+                else:
+                    args.append(None)
+                continue
+
+        if args and all(x is not None for x in args):
+            facts.append(frozenset(args))
+        else:
+            facts.append(None)
+    args = [arg for arg in facts if arg is not None]
+    return CNF(set(args))
```

Comment:
You can use a set comprehension directly rather than create a list and then turn it into a set: `{arg for arg in facts if arg is not None}` but... Why are you appending None to facts given that you then filter all Nones from facts?

It seems facts is defined as an empty list, then either has a frozenset appended or None (which is then filtered out). I'm not sure that the facts variable is needed at all.

I think that probably this whole function can be simplified substantially.

Reply:
Oh sorry, I misread the indentation of the loops. Forget what I said above

Reply:
Thanks. I simplified it.

Reply:
> Oh sorry, I misread the indentation of the loops. Forget what I said above

I don't know but your opinion seems genuine. I have changed the code and it seems fine.

Reply:
> It seems facts is defined as an empty list, then either has a frozenset appended or None (which is then filtered out). I'm not sure that the facts variable is needed at all.

Ok I understand now

Reply:
Maybe this:
```python
def _extract_all_facts(expr, symbol):
    if isinstance(symbol, Relational):
        symbols = (symbol, symbol.reversed)
    else:
        symbols = (symbol,)

    facts = set()
    for clause in expr.clauses:
        args = []
        for literal in clause:
            if isinstance(literal.lit, AppliedPredicate):
                if literal.lit.arg not in symbols:
                    break
                args.append(Literal(literal.lit.func, literal.is_Not))
        else:
            if args:
                facts.add(frozenset(args))

    return CNF(facts)
```
Note that the break here gives an early exit from the literal loop. There should be a meaningful comment explaining why that happens and no fact is added in that case...

Reply:
I had the exact changes. Will add a comment.